src/pgduckdb_utils.cpp

```diff
@@ -0,0 +1,63 @@
+#include "duckdb.hpp"
+
+extern "C" {
+#include "postgres.h"
+}
+
+#include "pgduckdb/pgduckdb_utils.hpp"
+
+namespace pgduckdb {
+
+template <typename T, typename FuncType, typename... FuncArgs>
+T
+PostgresFunctionGuard(FuncType postgres_function, FuncArgs... args) {
+	T return_value;
+	bool error = false;
+	MemoryContext ctx = CurrentMemoryContext;
+	ErrorData *edata = nullptr;
+	// clang-format off
+	PG_TRY();
+	{
+		return_value = postgres_function(args...);
+	}
+	PG_CATCH();
+	{
+		MemoryContextSwitchTo(ctx);
+		edata = CopyErrorData();
+		FlushErrorState();
+		error = true;
+	}
+	PG_END_TRY();
+	// clang-format on
+	if (error) {
+		throw duckdb::InternalException("(PGDuckDB/PostgresFunctionGuard) %s", edata->message);
+	}
+	return return_value;
+}
+
+template <typename FuncType, typename... FuncArgs>
+void
+PostgresFunctionGuard(FuncType postgres_function, FuncArgs... args) {
+	bool error = false;
+	MemoryContext ctx = CurrentMemoryContext;
+	ErrorData *edata = nullptr;
+	// clang-format off
+	PG_TRY();
+	{
+		postgres_function(args...);
+	}
+	PG_CATCH();
+	{
+		MemoryContextSwitchTo(ctx);
+		edata = CopyErrorData();
+		FlushErrorState();
+		error = true;
+	}
+	PG_END_TRY();
+	// clang-format on
+	if (error) {
+		throw duckdb::InternalException("(PGDuckDB/PostgresFunctionGuard) %s", edata->message);
```

Comment:
I think throwing an InternalException is acceptable for now, so that we can start using these guards in new code ASAP. But I think it would be great, if we would save not only the error message, but also the code/detail/context and maybe more in the `extra_data` of the `Exception`. Or probably easier: just store the whole `edata` (which we'd then probably have to allocate in `TopMemoryContext`).

That way we could convert such an `Exception` back to the equivalent `ereport` when surfacing the error to the user.